jsonrpc/debug_endpoint.go

```diff
@@ -0,0 +1,231 @@
+	EnableMemory     bool    `json:"enableMemory"`
+	DisableStack     bool    `json:"disableStack"`
+	DisableStorage   bool    `json:"disableStorage"`
+	EnableReturnData bool    `json:"enableReturnData"`
+	Timeout          *string `json:"timeout"`
+}
+
+func (d *Debug) TraceBlockByNumber(
+	blockNumber BlockNumber,
+	config *TraceConfig,
+) (interface{}, error) {
+	num, err := GetNumericBlockNumber(blockNumber, d.store)
+	if err != nil {
+		return nil, err
+	}
+
+	block, ok := d.store.GetBlockByNumber(num, true)
+	if !ok {
+		return nil, fmt.Errorf("block %d not found", num)
+	}
+
+	return d.traceBlock(block, config)
+}
+
+func (d *Debug) TraceBlockByHash(
+	blockHash types.Hash,
+	config *TraceConfig,
+) (interface{}, error) {
+	block, ok := d.store.GetBlockByHash(blockHash, true)
+	if !ok {
+		return nil, fmt.Errorf("block %s not found", blockHash)
+	}
+
+	return d.traceBlock(block, config)
+}
+
+func (d *Debug) TraceBlock(
+	input string,
+	config *TraceConfig,
+) (interface{}, error) {
+	blockByte, decodeErr := hex.DecodeHex(input)
+	if decodeErr != nil {
+		return nil, fmt.Errorf("unable to decode block, %w", decodeErr)
+	}
+
+	block := &types.Block{}
+	if err := block.UnmarshalRLP(blockByte); err != nil {
+		return nil, err
+	}
+
+	return d.traceBlock(block, config)
+}
+
+func (d *Debug) TraceTransaction(
+	txHash types.Hash,
+	config *TraceConfig,
+) (interface{}, error) {
+	tx, block := GetTxAndBlockByTxHash(txHash, d.store)
+	if tx == nil {
+		return nil, fmt.Errorf("tx %s not found", txHash.String())
+	}
+
+	if block.Number() == 0 {
+		return nil, ErrTraceGenesisBlock
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tracer, err := newTracer(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.store.TraceTxn(block, tx.Hash, tracer)
+}
+
+func (d *Debug) TraceCall(
+	arg *txnArgs,
+	filter BlockNumberOrHash,
+	config *TraceConfig,
+) (interface{}, error) {
+	header, err := GetHeaderFromBlockNumberOrHash(filter, d.store)
+	if err != nil {
+		return nil, ErrHeaderNotFound
+	}
+
+	tx, err := DecodeTxn(arg, d.store)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the caller didn't supply the gas limit in the message, then we set it to maximum possible => block gas limit
+	if tx.Gas == 0 {
+		tx.Gas = header.GasLimit
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tracer, err := newTracer(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.store.TraceCall(tx, header, tracer)
+}
+
+func (d *Debug) traceBlock(
+	block *types.Block,
+	config *TraceConfig,
+) (interface{}, error) {
+	if block.Number() == 0 {
+		return nil, ErrTraceGenesisBlock
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tracer, err := newTracer(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.store.TraceBlock(block, tracer)
+}
+
+// newTracer creates new tracer by config
+func newTracer(
+	ctx context.Context,
```

Comment:
Passed in context is not used, but I see it is correctly created and defer cancelled in all usages - but without interacting with it.

I don't think we need it, and the only cancellation context we need is the timeout done below inside the newTracer (`timeoutCtx`). Do you agree?

Reply:
Honestly this function doesn't need to take context because no one needs to stop. I will remove it

Reply:
Removed f3fcdf99bf024ceb8f9be26c5dcfb60a8c0750ad